### PR TITLE
User

### PR DESCRIPTION
**Pull request by nicole-a-tesla**

### CODE REVIEW

**Comment by ecmendenhall** [comment]
Try to squash out these WIP commits and any commits with failing builds before merging feature branches into `master`.

**Comment by ecmendenhall** [comment]
Taking the extra time and discipline to make sure every commit on master is working takes effort, but pays off in production.

File: meetuppizza/tests.py [hunk 1]
```diff
@@ -1,6 +1,23 @@
 from django.test import TestCase
+from django.contrib.auth.models import User
+from django.test import Client
+from meetuppizza.forms import RegistrationForm
+from django.contrib import auth
+
+import pdb
```
Comment (ecmendenhall) [inline meetuppizza/tests.py:7]: :scream_cat:

File: meetuppizza/tests.py [hunk 2]
```diff
@@ -1,6 +1,23 @@
...
+
 
 class Test(TestCase):
```
Comment (ecmendenhall) [inline meetuppizza/tests.py:10]: This is a bad name for this class. What is it actually testing? Do all the test methods belong together, or should they be split into several conceptually related classes?

File: meetuppizza/tests.py [hunk 3]
```diff
@@ -1,6 +1,23 @@
...
+  def setUp(self):
+    self.params =  {
+      'username':'Bjorn', 
+      'email':'bjorn@bjorn.com', 
+      'password1':'bjornbjorn', 
+      'password2':'bjornbjorn'
+    }
+
+    self.client = Client()
```
Comment (ecmendenhall) [inline meetuppizza/tests.py:19]: I thought this was already defined on instances of `django.test.TestCase`. Is this new in 1.9, maybe?

File: meetuppizza/tests.py [hunk 4]
```diff
@@ -8,3 +25,50 @@ def test_landing_page_is_there(self):
   def test_page_contains_pizza(self):
     response = self.client.get('/')
     self.assertContains(response, "pizza")
+
+  def test_signup_redirects(self):
```
Comment (ecmendenhall) [inline meetuppizza/tests.py:29]: Redirects to where?

File: meetuppizza/tests.py [hunk 5]
```diff
@@ -8,3 +25,50 @@ def test_landing_page_is_there(self):
...
+    response = self.client.post('/sign_up', self.params, follow=True)
+    self.assertRedirects(response, '/')
+
+  def test_user_is_created(self):
```
Comment (ecmendenhall) [inline meetuppizza/tests.py:33]: Make sure the names of your test methods include all the relevant what/when/where/why/how of the tests they describe. When is this user created?

File: meetuppizza/tests.py [hunk 6]
```diff
@@ -8,3 +25,50 @@ def test_landing_page_is_there(self):
...
+    self.client.post('/sign_up', self.params)
+    self.assertEqual(1, len(User.objects.all()))
+
+  def test_user_is_logged_in_after_signup(self):
```
Comment (ecmendenhall) [inline meetuppizza/tests.py:37]: :+1: this is a better name: it describes everything you're testing in the method body.

File: meetuppizza/tests.py [hunk 7]
```diff
@@ -8,3 +25,50 @@ def test_landing_page_is_there(self):
...
+    self.client.post('/sign_up', self.params)
+    user = User.objects.get(username='Bjorn')
+    self.assertFalse(user.is_anonymous())
+
+  def test_email_displayed_on_home_page(self):
```
Comment (ecmendenhall) [inline meetuppizza/tests.py:42]: Whose email?

File: meetuppizza/tests.py [hunk 8]
```diff
@@ -8,3 +25,50 @@ def test_landing_page_is_there(self):
...
+    self.client.post('/sign_up', self.params)
+    response = self.client.get('/')
+    self.assertContains(response, "bjorn@bjorn.com")
+
+  def test_user_log_out(self):
```
Comment (ecmendenhall) [inline meetuppizza/tests.py:47]: What's different after the user logs out?

File: meetuppizza/tests.py [hunk 9]
```diff
@@ -8,3 +25,50 @@ def test_landing_page_is_there(self):
...
+    self.client.post('/sign_up', self.params)
+    self.client.get('/sign_out')
+    user = auth.get_user(self.client)
+    self.assertTrue(user.is_anonymous())
+
+  def test_login(self):
```
Comment (ecmendenhall) [inline meetuppizza/tests.py:53]: What changes after a user logs in?

File: meetuppizza/tests.py [hunk 10]
```diff
@@ -8,3 +25,50 @@ def test_landing_page_is_there(self):
...
+    self.client.get('/sign_out')
+    login_params = {
+      'username':'Bjorn', 
+      'password':'bjornbjorn',     
+    }
+    self.client.post('/sign_in', login_params)
+    user = auth.get_user(self.client)
+    self.assertFalse(user.is_anonymous())
+
+  def test_invalid_login(self):
```
Comment (ecmendenhall) [inline meetuppizza/tests.py:64]: What happens after an invalid login?

File: meetuppizza/forms.py [hunk 11]
```diff
@@ -0,0 +1,10 @@
+from django import forms
+from django.contrib.auth.forms import UserCreationForm
+
+
+class RegistrationForm(UserCreationForm):
```
Comment (ecmendenhall) [inline meetuppizza/forms.py:5]: This form is totally testable, but I don't see any unit tests for it. :unamused:

How did it spring into existence without failing unit tests?

File: meetuppizza/views.py [hunk 12]
```diff
@@ -1,8 +1,52 @@
 from django.http import HttpResponse
 from django.template import loader
 from django.shortcuts import render
+from django.shortcuts import redirect
+from django.contrib.auth.models import User
+from django.contrib.auth import authenticate, login, logout
+from meetuppizza.forms import RegistrationForm
+import pdb
```
Comment (ecmendenhall) [inline meetuppizza/views.py:8]: :scream_cat:

File: meetuppizza/views.py [hunk 13]
```diff
@@ -1,8 +1,52 @@
...
 def index(request):
-  # template_name = "template/index.html"
-  # template = loader.get_template('templates/index.html')
-  return render(request, 'index.html')
\ No newline at end of file
+  return render(request, 'index.html')
+
+def sign_up(request):
+  if request.method == 'GET':
+    args = {}
```
Comment (ecmendenhall) [inline meetuppizza/views.py:15]: The thing called `args` here is usually called the template [context](https://docs.djangoproject.com/es/1.9/ref/templates/api/#django.template.Context).

File: templates/index.html [hunk 14]
```diff
@@ -2,6 +2,30 @@
 
 <link rel="stylesheet" type="text/css" href="{% static 'css/style.css' %}" >
 
-<div>
-  <h1>welcome to meetup.pizza</h1>
+
+<div class="welcome-header">
+
+  {% if not user.is_anonymous %}
```
Comment (ecmendenhall) [inline templates/index.html:8]: Is there a way to flip this logic instead of using an `if not`?

File: meetuppizza/tests.py (hunk 1 above)
Comment (nicole-a-tesla) [inline meetuppizza/tests.py:7]: :scream_cat:?

File: meetuppizza/tests.py (hunk 2 above)
Comment (nicole-a-tesla) [inline meetuppizza/tests.py:10]: Renamed and split up tests into better named classes

File: meetuppizza/tests.py (hunk 3 above)
Comment (nicole-a-tesla) [inline meetuppizza/tests.py:19]: Removed self.client = Client()

File: meetuppizza/tests.py (hunk 4 above)
Comment (nicole-a-tesla) [inline meetuppizza/tests.py:29]: renamed

File: meetuppizza/tests.py (hunk 5 above)
Comment (nicole-a-tesla) [inline meetuppizza/tests.py:33]: :+1:

File: meetuppizza/tests.py (hunk 6 above)
Comment (nicole-a-tesla) [inline meetuppizza/tests.py:37]: :+1:

File: meetuppizza/tests.py (hunk 7 above)
Comment (nicole-a-tesla) [inline meetuppizza/tests.py:42]: :+1:

File: meetuppizza/tests.py (hunk 8 above)
Comment (nicole-a-tesla) [inline meetuppizza/tests.py:47]: 🍦

File: meetuppizza/tests.py (hunk 9 above)
Comment (nicole-a-tesla) [inline meetuppizza/tests.py:53]: 🎉

File: meetuppizza/tests.py (hunk 10 above)
Comment (nicole-a-tesla) [inline meetuppizza/tests.py:64]: 💯

File: meetuppizza/forms.py (hunk 11 above)
Comment (nicole-a-tesla) [inline meetuppizza/forms.py:5]: 🚎

File: meetuppizza/views.py (hunk 12 above)
Comment (nicole-a-tesla) [inline meetuppizza/views.py:8]: 🍣

File: meetuppizza/views.py (hunk 13 above)
Comment (nicole-a-tesla) [inline meetuppizza/views.py:15]: changed

File: templates/index.html (hunk 14 above)
Comment (nicole-a-tesla) [inline templates/index.html:8]: there is!

**Pull request merged**